client/mm/libxc/bitget.go

```diff
@@ -0,0 +1,3133 @@
+
+	symbol      string
+	bookStrings *stringOrderbook // Primary: string orderbook (source of truth)
+	bookCache   *orderbook       // Secondary: cached uint64 orderbook for VWAP
+	bookDirty   atomic.Bool      // True when bookCache needs rebuild
+	updateQueue chan *bgtypes.BookUpdate
+
+	baseConversionFactor  uint64
+	quoteConversionFactor uint64
+	log                   dex.Logger
+
+	connectedChan chan bool
+
+	getSnapshot func() (*bgtypes.OrderbookSnapshot, error)
+
+	// Sequence tracking for data integrity
+	lastSeq                atomic.Int64 // Last processed sequence number
+	lastUpdateTs           atomic.Int64 // Last update timestamp (ms)
+	lastChecksumValidation atomic.Int64 // Last checksum validation time (ms)
+}
+
+// stringOrderbook maintains orderbook in original string format for checksum validation
+type stringOrderbook struct {
+	mtx  sync.RWMutex
+	bids [][]string // [price, quantity] sorted descending by price
+	asks [][]string // [price, quantity] sorted ascending by price
+}
+
+func newStringOrderbook() *stringOrderbook {
+	return &stringOrderbook{
+		bids: make([][]string, 0),
+		asks: make([][]string, 0),
+	}
+}
+
+// update applies changes to string orderbook
+func (sob *stringOrderbook) update(bids, asks [][]string) {
+	sob.mtx.Lock()
+	defer sob.mtx.Unlock()
+
+	// Update bids
+	for _, update := range bids {
+		if len(update) < 2 {
+			continue
+		}
+		price := update[0]
+		qty := update[1]
+
+		// Find existing price level
+		found := false
+		for i, level := range sob.bids {
+			if level[0] == price {
+				if qty == "0" {
+					// Remove
+					sob.bids = append(sob.bids[:i], sob.bids[i+1:]...)
+				} else {
+					// Update
+					sob.bids[i][1] = qty
+				}
+				found = true
+				break
+			}
+		}
+
+		// Insert new level if not found and qty != 0
+		if !found && qty != "0" {
+			sob.bids = sob.insertBid(price, qty)
+		}
+	}
+
+	// Update asks
+	for _, update := range asks {
+		if len(update) < 2 {
+			continue
+		}
+		price := update[0]
+		qty := update[1]
+
+		// Find existing price level
+		found := false
+		for i, level := range sob.asks {
+			if level[0] == price {
+				if qty == "0" {
+					// Remove
+					sob.asks = append(sob.asks[:i], sob.asks[i+1:]...)
+				} else {
+					// Update
+					sob.asks[i][1] = qty
+				}
+				found = true
+				break
+			}
+		}
+
+		// Insert new level if not found and qty != 0
+		if !found && qty != "0" {
+			sob.asks = sob.insertAsk(price, qty)
+		}
+	}
+}
+
+// insertBid inserts a bid maintaining descending price order
+func (sob *stringOrderbook) insertBid(price, qty string) [][]string {
+	priceFloat, _ := strconv.ParseFloat(price, 64)
+	newBid := []string{price, qty}
+
+	for i, level := range sob.bids {
+		levelPrice, _ := strconv.ParseFloat(level[0], 64)
+		if priceFloat > levelPrice {
+			// Insert here
+			return append(sob.bids[:i], append([][]string{newBid}, sob.bids[i:]...)...)
+		}
+	}
+	// Append at end
+	return append(sob.bids, newBid)
+}
```

Comment:
Seems really inefficient to parse everything every time rather than just converting once and saving as a float.

There's probably a better structure to save these in as well. Unsure atm though. I guess can optimize later.